### PR TITLE
feat: add orbax checkpointing support - closes #208 #211

### LINKED ISSUE #208: feat: add orbax checkpointing
**Is your feature request related to a problem? Please describe.**
Currently, training resumes are not fully deterministic and checkpointing the data pipeline is complex. The root cause is a **split state** between CPU and GPU:
1.  **CPU (Grain)**: Manages file reading and shuffling logic.
2.  **GPU (Sampler)**: Manages its own JAX `PRNGKey` and internal iteration counters (e.g., `batches_per_flow`).

If we only checkpoint Grain, the Sampler's RNG resets on resume, causing it to generate different images for the same flow field. Synchronizing these two independent states via manual serialization is fragile and error-prone.

**Describe the solution you'd like**
I propose a unified strategy that refines our randomness handling to enable simple, robust checkpointing via Orbax.

### 1. Refined Randomness: Grain as the Source of Truth
To make the Sampler restorable, we must make it **stateless** regarding the "source of entropy". 
- **Grain Path**: We add a `RandomMapTransform` to the Grain pipeline. For every record, Grain provides a deterministic **integer seed** (`jax_seed`).
- **Sampler Path**: The Sampler converts the integer into a JAX key (`jax.random.PRNGKey(jax_seed)`) and then uses `jax.random.fold_in(key, rep_idx)` to derive unique noise for each variation. 
- **Legacy Path**: To maintain backward compatibility, the Sampler falls back to an internal `_rng` if no seed is provided by the scheduler.

### 2. Modern Orbax Checkpointing
By moving the RNG "anchor" to Grain, we can use the **Modern Orbax Args-based API** to save the entire pipeline state in one atomic transaction.

#### Saving
```python
mngr.save(
    step=global_step,
    args=ocp.args.Composite(
        # Handles binary Grain iterator state (records, shuffle position)
        data=grain.PyGrainCheckpointSave(grain_iterator),
        # Handles Sampler metadata (step counter, legacy _rng) as a PyTree
        sampler=ocp.args.StandardSave({"step": sampler._step, "_rng": sampler._rng}),
    )
)
```

#### Restoring
```python
restored = mngr.restore(
    step=ckpt_step,
    args=ocp.args.Composite(
        data=grain.PyGrainCheckpointRestore(grain_iterator),
        sampler=ocp.args.StandardRestore({"step": 0, "_rng": initial_rng}),
    )
)
# The sampler is restored perfectly because both the record and the counter 
# (which hashes the noise) are synchronized.
sampler._step = restored.sampler["step"]
sampler._rng = restored.sampler["_rng"]
```

**Why this is necessary**
This change turns the randomness problem from a "state synchronization" problem into a "derivation" problem. By hashing the Grain-provided seed with the Sampler-provided step, we guarantee that **variation #3 of flow #401** is identical every time, regardless of whether it's a first run or a resume from checkpoint.

**Additional context**
- This follows the official [Grain-Orbax integration guide](https://google-grain.readthedocs.io/en/latest/tutorials/data_loader_tutorial.html#integration-with-orbax).

### PR DESCRIPTION
**Pull request by fbanelli**

### CODE REVIEW

File: tests/test_grain_compatibility/test_checkpointing.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    config = {
+        "scheduler_class": ".mat",
+        "batch_size": 2,
+        "include_images": True,
+        "flow_fields_per_batch": 2,
+        # We need a file list for real data sources
+        "file_list": ["dummy.mat"],
+        "image_shape": (64, 64),
+        "seed": 42
+    }
```
[the remark this answers is not in the document]
Comment (fbanelli) [reply tests/test_grain_compatibility/test_checkpointing.py:141]: if include_images=True those keys aren't required

File: docs/checkpointing.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
+# Checkpointing and Reproducibility in SynthPix 💾
+
+SynthPix provides a robust, bit-perfect checkpointing system built on top of [Orbax](https://orbax.readthedocs.io/). Developing data-hungry PIV algorithms requires large-scale training, and being able to pause/resume with guaranteed deterministic results is critical.
+
+## Core Concepts
+
+### Bit-Perfect Reproducibility
+SynthPix guarantees that the images generated after a restore are **identical** to those that would have been generated if the run had never stopped. This includes:
+1.  **Data Order**: Ensuring exactly the same flow field records are loaded from disk in the same order.
```
Comment (antonioterpin) [inline docs/checkpointing.md]: Data order

File: docs/checkpointing.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+2.  **Randomness**: Synchronizing the JAX RNG state on the GPU with the data loader's position on the CPU.
+3.  **Variation Logic**: Preserving repetition counters (e.g., when `batches_per_flow_batch > 1`).
```
Comment (antonioterpin) [inline docs/checkpointing.md]: Variation logic

File: docs/checkpointing.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+### Unified State
+A SynthPix pipeline has two main state components:
+-   **CPU State (Grain)**: Manages file pointers, shuffling logic, and the deterministic `jax_seed` passed to the GPU.
+-   **GPU State (Sampler)**: Manages internal iteration counters and (optionally) its own JAX RNG key for legacy paths.
+
+Orbax acts as the atomic transaction layer that saves both states simultaneously.
```
Comment (antonioterpin) [inline docs/checkpointing.md]: Orbax acts as the transaction layer that saves both states atomically.

File: docs/checkpointing.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
+# Checkpointing and Reproducibility in SynthPix 💾
```
Comment (antonioterpin) [inline docs/checkpointing.md]: Checkpointing and reproducibility in Synthpix

Comment (antonioterpin) [reply docs/checkpointing.md]: Similar comment below, please adhere to this capitalization

File: docs/checkpointing.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+The `Sampler` base class is the orchestrator for state. It implements `get_state()` and `set_state()`, which:
+1.  Captures/Restores internal sampler metadata (e.g., `_step`, `_batches_generated`).
+2.  Recursively calls the underlying `Scheduler` state methods.
+
+When `synthpix.save_checkpoint` is called, it:
+1.  Retrieves the sampler's dictionary state.
+2.  Retrieves the binary state of the Grain iterator (`grain.PyGrainCheckpointSave`).
+3.  Packs them into a `Composite` Orbax argument for atomic saving.
+
+### Randomness Derivation
```
Comment (antonioterpin) [inline docs/checkpointing.md]: Randomness, drop "Derivation"

File: docs/checkpointing.md [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+To ensure restorable randomness, SynthPix uses a **derivation** strategy:
```
Comment (antonioterpin) [inline docs/checkpointing.md]: To ensure deterministic randomness

File: docs/examples/checkpointing_example.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    # Initialize sampler with patched config dict
+    sampler = synthpix.make(config)
+    print(f"Sampler initialized with {type(sampler).__name__}")
+
+    # Run for a few steps
+    # We simulate a "training loop" here
+    for i, batch in enumerate(sampler):
+        print(f"[Run 1] Step {i}: Batch files: {batch.files}")
+
+        # Save checkpoint at step 2
```
Comment (antonioterpin) [inline docs/examples/checkpointing_example.py:71]: modify the example so that it is apparent that we have the option to automatically delete the old checkpoints if specified so

Comment (antonioterpin) [reply docs/examples/checkpointing_example.py:71]: Would be cool would be to give the option to regularly downsample the checkpoints.

Suppose one has checkpoints with downsample factor 3:
0 1 2 3 4 5
-> 0 3 5
-> 0 3 5 6
-> 0 3 5 6 7
-> 0 3 5 6 7 8
-> 0 6 8
-> 0 6 8 9 ....

Note sure if it is too important, but seems reasonable. I mean there are arbitrarily many options so probably not worth it.

File: docs/tutorials.md [hunk 9]
```diff
@@ -1,4 +1,7 @@
-# More examples 🌊
+# Tutorials and Feature Guides 🌊
+
+## Core Features
```
Comment (antonioterpin) [inline docs/tutorials.md:3]: Can you open issues for a few other tutorials that would make sense to add?

File: src/synthpix/sampler/real.py [hunk 10]
```diff
@@ -62,6 +64,24 @@ def _get_next(self) -> SynthpixBatch:
         )
         return batch
 
+    @property
```
Comment (antonioterpin) [inline src/synthpix/sampler/real.py]: You should not need to redefine getters and setters. You can inherit them

File: src/synthpix/sampler/synthetic.py [hunk 11]
```diff
@@ -547,8 +576,122 @@ def _get_next(self) -> SynthpixBatch:
...
+    @property
+    def state(self) -> dict[str, Any]:
+        """Returns the state of the sampler for checkpointing.
+
+        Returns:
+            A dictionary containing the sampler state.
+        """
+        state_dict = super().state
+
+        # Encode files_scheduler as uint8 array (valid for Orbax/TensorStore)
```
Comment (antonioterpin) [inline src/synthpix/sampler/synthetic.py]: define a function?

as_uint8(...)

**Review by antonioterpin — CHANGES_REQUESTED** [review]

File: docs/examples/checkpointing_example.py (hunk 8 above)
Comment (fbanelli) [reply docs/examples/checkpointing_example.py:71]: it's a bit arbitrary, if needed we could implement it but I wouldn't ship it out of the box

**Review by antonioterpin — APPROVED** [review]
